### PR TITLE
Add converge function

### PR DESCRIPTION
**Pull request by win0err**
## Converge

Accepts a converging function and a list of branching functions and returns a new function. When invoked, this new function is applied to some arguments, and each branching function is applied to those same arguments. The results of each branching function are passed as arguments to the converging function to produce the return value.

```php
<?php
converge('intdiv', ['max', 'min'])([2, 3, 4, 5, 6]); // 3
```

[JS ramda analog](https://ramdajs.com/docs/#converge)

### CODE REVIEW

**Comment by Erikvv** [comment]
What is the reason to do this instead of equivalent code which is also functional:

```
function middle(array $data) {
    intdiv(max($data), min($data));
}

middle([2, 3, 4, 5, 6]);
```

Converge would be useful if you don't know which functions to apply beforehand but I can't think of such case. 

Do you have a good usecase?

**Comment by win0err** [comment]
> What is the reason to do this instead of equivalent code which is also functional:
> 
> ```
> function middle(array $data) {
>     return intdiv(max($data), min($data));
> }
> 
> middle([2, 3, 4, 5, 6]);
> ```
> 
> Converge would be useful if you don't know which functions to apply beforehand but I can't think of such case.
> 
> Do you have a good usecase?
```php
<?php
// sma() function is taken from here: 
// https://github.com/markrogoyski/math-php/blob/master/src/Statistics/Average.php#L650
$sma = curry_n(2, flip('sma'));

$processingFunction = fn (...$args) => array_map(curry('implode')(' → '), $args);
$getChart = converge(
    $processingFunction,
    [$sma(1), $sma(2), $sma(4), $sma(8)]
);
// vs.
$getChart2 = function ($data) use ($sma) {
    return [
        implode(' → ', $sma(1)($data)),
        implode(' → ', $sma(2)($data)),
        implode(' → ', $sma(4)($data)),
        implode(' → ', $sma(8)($data)),
    ];
};

$chart = $getChart([1, 2, 3, 4, 5, 6, 7, 8, 9]);
```
Cons of `converge` is a more functional approach. It's easy to add another function like `$sma(42)` to the list of branching functions.

File: src/Functional/Converge.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * @param callable $convergingFunction
+ * @param callable[] $branchingFunctions
+ * @return callable a flipped version of the given function
+ */
+function converge($convergingFunction, array $branchingFunctions)
+{
+    return function (...$values) use ($convergingFunction, $branchingFunctions) {
+        $result = [];
+
+        foreach ($branchingFunctions as $f) {
```
Comment (lstrojny) [inline src/Functional/Converge.php]: `as $branchingFunction` for clarity?

File: tests/Functional/ConvergeTest.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        $strangeFunction = converge(
+            function ($u, $l) {
+                return $u . $l;
+            },
+            [
+                'strtoupper',
+                'strtolower',
+            ]
+        );
+        $this->assertEquals('FUNCTIONAL PROGRAMMINGfunctional programming', $strangeFunction('Functional Programming'));
```
Comment (lstrojny) [inline tests/Functional/ConvergeTest.php]: assertSame()

File: src/Functional/Converge.php (hunk 1 above)
Comment (win0err) [reply src/Functional/Converge.php]: Fixed

File: tests/Functional/ConvergeTest.php (hunk 2 above)
Comment (win0err) [reply tests/Functional/ConvergeTest.php]: Fixed

**Comment by lstrojny** [comment]
Awesome, thank you!

**Pull request merged**